uber/models/group.py

```diff
@@ -205,10 +201,116 @@ def total_cost(self):
     @property
     def amount_unpaid(self):
         if self.registered:
-            return max(0, self.cost - self.amount_paid)
+            return max(0, ((self.cost * 100) - self.amount_paid) / 100)
         else:
             return self.total_cost
 
+    @hybrid_property
+    def amount_paid(self):
+        return sum([item.amount for item in self.receipt_items if item.txn_type == c.PAYMENT])
+
+    @amount_paid.expression
+    def amount_paid(cls):
+        from uber.models import ReceiptItem
+
+        return select([func.sum(ReceiptItem.amount)]
+                      ).where(and_(ReceiptItem.group_id == cls.id,
+                                   ReceiptItem.txn_type == c.PAYMENT)).label('amount_paid')
+
+    @hybrid_property
+    def amount_refunded(self):
+        return sum([item.amount for item in self.receipt_items if item.txn_type == c.REFUND])
+
+    @amount_refunded.expression
+    def amount_refunded(cls):
+        from uber.models import ReceiptItem
+
+        return select([func.sum(ReceiptItem.amount)]
+                      ).where(and_(ReceiptItem.group_id == cls.id,
+                                   ReceiptItem.txn_type == c.REFUND)).label('amount_refunded')
+
+    def balance_by_item_type(self, item_type):
+        """
+        Return a sum of all the receipt item payments, minus the refunds, for this model by item type
+        """
+        return sum([amt for type, amt in self.itemized_payments if type == item_type]) \
+               - sum([amt for type, amt in self.itemized_refunds if type == item_type])
+
+    @property
+    def itemized_payments(self):
+        return [(item.item_type, item.amount) for item in self.receipt_items if item.txn_type == c.PAYMENT]
+
+    @property
+    def itemized_refunds(self):
+        return [(item.item_type, item.amount) for item in self.receipt_items if item.txn_type == c.REFUND]
+
+    @receipt_item
+    def badge_cost_receipt_item(self):
+        if not self.badge_cost:
+            return None, None, None
+
+        item_type = c.BADGE
+        badge_balance = self.balance_by_item_type(item_type)
+        paid_badges = badge_balance / self.new_badge_cost
+
+        return (self.badge_cost * 100) - badge_balance, \
+               "{} badge{} (${} each)".format(int(self.badges_purchased - paid_badges),
+                                             "s" if self.badges_purchased > 1 else "",
+                                             self.new_badge_cost), item_type
+
+    @receipt_item
+    def table_receipt_item_1st(self):
```

Comment:
It's not ideal that there's four hard-coded receipt items for each table cost, when you could technically have more than 4 unique table prices. However, you can override table_receipt_item_more and add more receipt items if you need, so I'm not too concerned about this and I had no good ideas on how to make this more flexible.

By the way, the function names are the way they are so that the order of receipt items is nicer-looking.

Reply:
Hrm, yeah that doesn't seem ideal, but the couple of "better" solutions I can think of involve a confusing amount of automatic code generation, e.g. we could write a function-generating-function and then iterate over ``c.TABLE_PRICES`` and call that function and attach the results to ``Root``.  Since we current never sell more than 4 tables, this approach seems fine, and we can re-evaluate if/when that ever changes.